src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,119 @@
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.mapper.DocValueFetcher;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.ValueFetcher;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.fetch.FetchContext;
+import org.opensearch.search.fetch.FetchSubPhase;
+import org.opensearch.search.fetch.FetchSubPhaseProcessor;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.BYTES_PER_KILOBYTES;
+
+
+/**
+ * Fetch sub phase which pull data from doc values.
+ * and fulfill the value into source map
+ */
+public class KNNFetchSubPhase implements FetchSubPhase {
+    private static Logger logger = LogManager.getLogger(KNNFetchSubPhase.class);
+
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) throws IOException {
+        if (!KNNSettings.isKNNSyntheticEnabled(fetchContext.getIndexName())) {
+            return null;
+        }
+        MapperService mapperService = fetchContext.mapperService();
+
+        List<DocValueField> fields = new ArrayList<>();
+        for (MappedFieldType mappedFieldType : mapperService.fieldTypes()) {
+            if (mappedFieldType != null && mappedFieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType) {
+                String fieldName = mappedFieldType.name();
+                ValueFetcher fetcher = new DocValueFetcher(mappedFieldType.docValueFormat(null, null),
+                        fetchContext.searchLookup().doc().getForField(mappedFieldType));
+                fields.add(new DocValueField(fieldName, fetcher));
+            }
+        }
+        return new KNNFetchSubPhaseProcessor(fetchContext, fields);
+    }
+
+    @AllArgsConstructor
+    @Getter
+    class KNNFetchSubPhaseProcessor implements FetchSubPhaseProcessor {
+
+        private final FetchContext fetchContext;
+        private final List<DocValueField> fields;
+
+        @Override
+        public void setNextReader(LeafReaderContext readerContext) throws IOException {
+            for (DocValueField f : fields) {
+                f.fetcher.setNextReader(readerContext);
+            }
+        }
+
+        @Override
+        public void process(HitContext hitContext) throws IOException {
+            MapperService mapperService = fetchContext.mapperService();
+            final boolean hasNested = mapperService.hasNested();
+            SearchHit hit = hitContext.hit();
+            Map<String, Object> maps = hit.getSourceAsMap();
+            if (maps == null) {
+                //when source is disabled, return
+                return;
+            }
+
+            if (hasNested) {
+                //TODO handle nested field
```

Comment:
Is the nested case difficult to handle?

Reply:
yes, it is.
in nested field, subPhase have to find inner hits doc values, and use it to synthetic the source map to parent doc source field. so, I think I can make an other pr for the nested field.

Reply:
Makes sense. Should this gracefully fail though by either throwing an exception or returning until the next PR? I worry that we release it and then do not get nested fix in in time

Reply:
LGTM  at https://github.com/opensearch-project/k-NN/pull/1571/commits/3f119b83d2017b28897f30d82c7f73305a5300a0 it would throw `UnsupportedOperationException`